### PR TITLE
Validate verification code input before sending it

### PR DESCRIPTION
**Pull request by marlam**
This makes sure that the verification code only contains characters a-z A-Z 0-9, and is of limited length.

Otherwise it is possible for a user to cause arbitrary data to be sent to the server, including invalid JSON. This could be a problem if the JSON parser on the receiving end has bugs.

For example, code input 'A"}BBBBBBBB...' would lead to JSON data '{"session_id":"SESSIONID","pin":"A"}BBBBBBBB..."}' to be sent to the server.

### CODE REVIEW

**Comment by mrvanes** [comment]
When using pamtester to test the PR, I get a segmentation fault on and empty code response:
```
Enter verification code: <enter>
Segmentation fault
```

It would be nice if you could extend the validation to the username input and return a `permission denied` on invalid usernames:
```
$ sudo pamtester weblogin '"a}aaa"' authenticate
Server error!
pamtester: System error
```

**Comment by marlam** [comment]
I fixed the segfault - I overlooked that tty_input() returns NULL on empty input.

This can also bite the group input, I opened PR #104 for this.

I will extent this to cover user names in a separate PR, because that requires a small extension of `input_is_safe()`.

File: src/tty.c [hunk 1]
```diff
@@ -77,6 +77,35 @@ char *tty_input(pam_handle_t *pamh, const char *text, int echo_code)
 	return ret;
 }
 
+int input_is_safe(const char *input, size_t max_length)
+{
+	size_t length = strlen(input);
```
Comment (baszoetekouw) [inline src/tty.c:82]: This is not safe.  If input isn't terminated by a `\0`, strlen() will run into unrelated memory.  This is unlikely but could possibly occur.

The better solution is to use `strnlen()` and then make sure that the string is 0-terminated.  Or use `strndup()` to make a local copy of the string.

File: src/tty.c [hunk 2]
```diff
@@ -77,6 +77,35 @@ char *tty_input(pam_handle_t *pamh, const char *text, int echo_code)
 	return ret;
 }
 
+int input_is_safe(const char *input, size_t max_length)
```
Comment (baszoetekouw) [inline src/tty.c:80]: If you intend to return a true/false value, please use an explicit `bool` value for that.

**Review by baszoetekouw — CHANGES_REQUESTED** [review]
Hi Martin!
Thanks for this PR.  Looks good in principle, but I think we should be a bit more careful with the string manipulations (see comments).

**Review by baszoetekouw — APPROVED** [review]

**Pull request merged**